pkg/client/clientpool.go

```diff
@@ -171,15 +180,60 @@ func (p *ClientPool) startWorker(i int) {
 
 		now := time.Now()
 
-		var err error = nil
-		var resp *nebula.ResultSet = nil
-		for retry := p.retry; retry > 0; retry-- {
+		exp := backoff.NewExponentialBackOff()
+		exp.InitialInterval = DefaultRetryInitialInterval
+		exp.RandomizationFactor = DefaultRetryRandomizationFactor
+		exp.Multiplier = DefaultRetryMultiplier
+		exp.MaxInterval = DefaultRetryMaxInterval
+		exp.MaxElapsedTime = DefaultRetryMaxElapsedTime
+
+		var (
+			err   error
+			resp  *nebula.ResultSet
+			retry = p.retry
+		)
+
+		// There are three cases of retry
+		// * Case 1: retry no more
+		// * Case 2. retry as much as possible
+		// * Case 3: retry with limit times
+		_ = backoff.Retry(func() error {
 			resp, err = p.Sessions[i].Execute(data.Stmt)
 			if err == nil && resp.IsSucceed() {
-				break
+				return nil
 			}
-			time.Sleep(1 * time.Second)
-		}
+			retryErr := err
+			if resp != nil {
+				errorCode, errorMsg := resp.GetErrorCode(), resp.GetErrorMsg()
+				retryErr = fmt.Errorf("%d:%s", errorCode, errorMsg)
+
+				// Case 1: retry no more
+				var isPermanentError = true
+				switch errorCode {
+				case nebula.ErrorCode_E_SYNTAX_ERROR:
+				case nebula.ErrorCode_E_SEMANTIC_ERROR:
+				default:
+					isPermanentError = false
+				}
+				if isPermanentError {
+					// stop the retry
+					return backoff.Permanent(retryErr)
+				}
+
+				// Case 2. retry as much as possible
+				if strings.Contains(errorMsg, "raft buffer is full") {
```

Comment:
Is there a error code to represent the raft buffer error? this comparison will fail when the error message updated. better to compare the error code.

Reply:
Yes, I do think so too. But there is not an error code represent it.

@jievince says that the error code structure will not be adjust in version 3.4.

https://github.com/vesoft-inc/nebula/blob/1aee9626ec6dd93a6c593cf0c953d72db7089ffe/src/graph/executor/StorageAccessExecutor.h#L127-L129

https://github.com/vesoft-inc/nebula/blob/1aee9626ec6dd93a6c593cf0c953d72db7089ffe/src/graph/executor/algo/ShortestPathBase.cpp#L144-L146